src/neo/SmartContract/Native/NotaryContract.cs

```diff
@@ -0,0 +1,325 @@
+            if (deposit != null && till < deposit.Till) throw new Exception(string.Format("`till` shouldn't be less then the previous value {0}", deposit.Till));
+            if (deposit is null)
+            {
+                if ((long)amount < 2 * GetNotaryServiceFeePerKey(engine.Snapshot)) throw new Exception(string.Format("first deposit can not be less then {0}, got {1}", 2 * GetNotaryServiceFeePerKey(engine.Snapshot), amount));
+                deposit = new Deposit() { Amount = 0, Till = 0 };
+                if (!allowedChangeTill) till = currentHeight + DefaultDepositDeltaTill;
+            }
+            else if (!allowedChangeTill) till = deposit.Till;
+            deposit.Amount += amount;
+            deposit.Till = till;
+            PutDepositFor(engine, to, deposit);
+        }
+
+        /// <summary>
+        /// Lock asset until the specified height is unlocked
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="addr">Account</param>
+        /// <param name="till">specified height</param>
+        /// <returns>result</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        public bool LockDepositUntil(ApplicationEngine engine, UInt160 addr, uint till)
+        {
+            if (!engine.CheckWitnessInternal(addr)) return false;
+            if (till < Ledger.CurrentIndex(engine.Snapshot)) return false;
+            Deposit deposit = GetDepositFor(engine.Snapshot, addr);
+            if (deposit is null) return false;
+            if (till < deposit.Till) return false;
+            deposit.Till = till;
+            PutDepositFor(engine, addr, deposit);
+            return true;
+        }
+
+        /// <summary>
+        /// Withdraw sends all deposited GAS for "from" address to "to" address.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="from">From Account</param>
+        /// <param name="to">To Account</param>
+        /// <returns>void</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        private async ContractTask<bool> Withdraw(ApplicationEngine engine, UInt160 from, UInt160 to)
+        {
+            if (!engine.CheckWitnessInternal(from)) throw new InvalidOperationException(string.Format("Failed to check witness for {0}", from.ToString()));
+            var receive = to is null ? from : to;
+            Deposit deposit = GetDepositFor(engine.Snapshot, from);
+            if (deposit is null) throw new InvalidOperationException(string.Format("Deposit of {0} is null", from.ToString()));
+            if (Ledger.CurrentIndex(engine.Snapshot) < deposit.Till) throw new InvalidOperationException(string.Format("Can't withdraw before {0}", deposit.Till));
+            await GAS.Burn(engine, Hash, deposit.Amount);
+            await GAS.Mint(engine, receive, deposit.Amount, true);
```

Comment:
It should be able to perform withdrawal in one step. We'd better not to misuse Burn and Mint, let's squash these two methods.